### ISSUE TITLE
Add support for Fedora, and/or Flatpak for all distros

### ISSUE THREAD

**Issue by heyakyra** [OpenBazaar/openbazaar-desktop#383]
Fedora users will not be able to use the "Linux" download provided. It is a bit misleading. https://github.com/OpenBazaar/openbazaar.org/issues/64

There should be .rpm files for them, and Flatpak could be used to support all GNU/Linux distributions.

**Comment by hoffmabc** [comment]
I think we specify that we only officially support Ubuntu, but I can see how this is misleading. I don't know if we will add RPMs but we'll put this in the queue.

**Comment by hoffmabc** [comment]
Oh also I think we're looking at Snap already as well.

**Comment by thekyriarchy** [comment]
Oh cool. Snap is unfortunately an Ubuntu project and they're known to not always play nice with the rest of the distro community (duplicate effortts so that they have full control). Comparing the adoption rate of each speaks for itself i'd say: https://kamikazow.wordpress.com/2017/02/09/adoption-of-flatpak-vs-snap/

**Comment by hoffmabc** [comment]
We have two Canonical team members in our Slack @elopio and @popey who might have some thoughts on this as well.

**Comment by popey** [comment]
🔮 Spell of summoning worked!  ✨ 

I can appreciate that people have different opinions on Canonical and snaps. The fact is we're working hard to get first class snap support in a lot of distros right now. That's not an easy job, especially given the various rules and policies we have to adhere to across them all. Zygmunt Krynicki - our lead developer in this area - wrote some thoughts on it recently. Worth a read for a fuller picture. https://new.zygoon.pl/post/state-of-snapd-support-across-distros/ 

Personally I have snaps running on Ubuntu 14.04, 16.04, 16.10, (32 and 64-bit), Ubuntu Core on ARM (Raspberry Pi), OpenSUSE Leap, KDE Neon, Linux Mint, Zorin OS, Peppermint OS and Antergos (Arch). Some of those are of course Ubuntu derivatives, but in total they all cover quite a huge chunk of the Linux-using market. More so if you consider that Ubuntu alone has the lions share of the cloud and server space.

Of course none of these packaging systems are mutually exclusive. It's entirely possible to have snaps, RPM, Deb and Appimage. It's just a case of where your priorities lie, and how much of your finite engineering effort you're willing to expend to get there. 

With snaps, once you have done the up-front work to get a working snap package, we have the build and store infrastructure to easily distribute them on every update to an audience of millions, which for many is a compelling argument.

**Comment by elopio** [comment]
@thekyriarchy we come in peace. After helping to prepare the [openbazaar-server snap](https://github.com/OpenBazaar/openbazaar-go/blob/master/snapcraft.yaml) now I think that the project is great and I just want it to be easy and safe to install everywhere. I will certainly appreciate any efforts from the Fedora community to help there. From the Ubuntu community we are proposing snaps, and I honestly think that it would be great for our users and for the openbazaar project to have continuous delivery to the snappy store, no foul play there. 
BTW, @tyler-smith, I was working on a fancier way to get go 1.7 during the build, I'm sending a PR your way :) 

In the end the project will choose what's better for them, and as popey said, they can even decide on more than one distribution method. No harm on trying any of those methods.

**Comment by thekyriarchy** [comment]
thanks for the responses and i'm glad it's being worked on in any capacity! 

*as an aside*: Thanks for link on details on snap packaging. My aversion to canonical and its founder are mostly stemming from their CLA and other things that make me apprehensive. This doesn't extend to their employees, as hearing the first-hand experiences of several of them has contributed to my disaffection. 

As far as this bug report goes, I'll be just as satisfied if snap can get full support on Fedora and I can use OpenBazaar that way as if .rpm or Flatpak builds are made available, despite my preferences generally.

**Comment by tyler-smith** [comment]
@elopio's PR is merged and seems to be [building correctly.](https://launchpad.net/~openbazaar/+snap/openbazaar)

@thekyriarchy How do you feel about Docker?

`docker run --rm -v ~/ob:/var/openbazaar -P ob1company/openbazaar`

**Comment by elopio** [comment]
@tyler-smith almost correctly:
`Store upload failed: You must register openbazaar-server name for 16 before uploading, please register it and try again.`

The name in the yaml is `openbazaar-server`, and the name you registered is `openbazaar`. You just have to choose one and either update the yaml, or register `openbazaar-server` and use it in launchpad.

**Comment by thekyriarchy** [comment]
@tyler-smith oh cool! i'll try that out (though to answer the question, I prefer rocket 😉)

**Comment by tyler-smith** [comment]
@elopio Ah, I see. I'll fix that.

**Comment by probonopd** [comment]
> Of course none of these packaging systems are mutually exclusive. It's entirely possible to have snaps, RPM, Deb and Appimage. It's just a case of where your priorities lie, and how much of your finite engineering effort you're willing to expend to get there.

Well said. Providing an [AppImage](http://appimage.org/) would have, among others, these advantages:
- Works for most Linux distributions (including Ubuntu, Fedora, openSUSE, CentOS, elementaryOS, Linux Mint, and others)
- One app = one file = super simple for users: just download one AppImage file, [make it executable](http://discourse.appimage.org/t/how-to-make-an-appimage-executable/80), and run
- No unpacking or installation necessary
- No root needed
- No system libraries changed
- Just one format for all major distributions
- Works out of the box, no installation of runtimes needed (unlike Snap and Flatpak)
- Optional desktop integration with `appimaged`
- Optional binary delta updates, e.g., for continuous builds (only download the binary diff) using AppImageUpdate
- Can optionally GPG2-sign your AppImages (inside the file)
- Works on Live ISOs (unlike Flatpak)
- Can use the same AppImages when dual-booting multiple distributions

[Here is an overview](https://github.com/probonopd/AppImageKit/wiki/AppImages) of projects that are already distributing upstream-provided, official AppImages. Many of them Electron-based.

**Comment by snaggen** [comment]
While AppImages are simple and functional, they are quite limited compared to Flatpaks. 
* Flatpaks requires a lot less space, since they have common runtimes. They are also built on ostree, to further remove duplicate files taking up space. 
* Flatpaks have great desktop integration, making them super simple to use for users. Basically you install a flatpak with a click in the browser (https://youtu.be/fK4plb8aVhM). Once a flatpak is installed, you can launch from your desktop launcher like any other application. Gnome Software will also allow you to manage and upgrade your application. 
* Flatpaks are secure. You can sandbox the application, and limit what access it have to the outside world. You then have portals to let the user grant specific access, like for saving a file or open a link in the users default browser. 

So basically, A flatpak provides an easier user experience, takes up less space and provides far better security.

**Comment by probonopd** [comment]
Hi @snaggen, it's not as black-and-white:
* An AppImage runtime does not need to be installed on the system before you can run an AppImage, unlike the other solutions. This means it will run immediately on Ubuntu and other distributions. This is a biggie, at least for me
* Desktop integration can be done e.g., using the `desktopintegration` bash script or the optional `appimaged` daemon
* AppImage does not litter files around the filesystem, which means you can manage them yourself rather than needing a package manager
* The same AppImage can be run when you multi-boot different distributions
* AppImages are easy to archive

**Comment by HoboPrimate** [comment]
Any update on this? Have been waiting for a way to install openbazaar on Fedora since 1.0 days...

**Comment by jjeffryes** [comment]
We're not actively working on this, but if other people want to chip in, we'd welcome the help.

**Comment by snaggen** [comment]
@probonopd AppImages are great in many ways, don't get me wrong. However, the problem with AppImages is that it only almost solves one problem of what flatpak solves. The lack of runtime means that the developer that packages the application doesn't have a runtime to target and package for, but still have to target separate distributions. From their own docs: 

> For example, if you are targeting Ubuntu, Fedora, and openSUSE, then you need to gather all libraries and other dependencies that your app requires to run that are not part of Ubuntu, Fedora, and openSUSE

Which means that developer still have to know what libraries are part of which distribution, or just package EVERYTHING, including really basic libs from /lib like libssl and such. So even if your AppImage runs on a Ubuntu, it is only likely to run on other distributions there is no guarantee. The approach taken by flatpak, is that you compile your software against a SDK, and only package what is missing from that. If your application works on one distribution, it will then run on all since the runtime is guaranteed to have everything it needs.

**Comment by probonopd** [comment]
@snaggen yes, AppImages make it harder for the developer in order to make it easier for the user. Just download, chmod +x, done.

> Which means that developer still have to know what libraries are part of which distribution, or just package EVERYTHING, including really basic libs from /lib like libssl and such.

No. While a developer _can_ decide what to bundle, the general rule of thumb is to bundle everything _but core libraries that can be assumed to be part of each target system in a recent enough version_ like [these](https://github.com/AppImage/AppImages/blob/master/excludelist).

From my personal experience, I've never gotten to work Flatpak on Live ISOs (and that's all I personally ever use).

**Comment by elopio** [comment]
I'm still happy to help with the snap. As far as I remember, what was left was a mismatch between the registered name and the name declared in the snap. Just ping me if you need a hand with that, or anything else.

**Comment by hoffmabc** [comment]
This next release will have RPMs and possibly flatpak support as well.

**Comment by hoffmabc** [comment]
Ok running into an issue with flatpak builds because it requires Ubuntu 16 and up. Travis only supports Xenial through docker. This would require more work so will probably punt on flatpak for the moment.

**Comment by snaggen** [comment]
Have you tested the ppa https://launchpad.net/~alexlarsson/+archive/ubuntu/flatpak
I'm not sure about the ppa and what versions it supports, but if you need some other version that the ppa build system can support, I'm sure Alex could try to setup a build for that.

**Comment by hoffmabc** [comment]
I'll check it out. Let me see what I get.

**Comment by hoffmabc** [comment]
I didn't use the +archive link but the regular one before, but the one you suggested @snaggen gave a 404.

```
W: Failed to fetch https://launchpad.net/~alexlarsson/+archive/ubuntu/flatpak/dists/trusty/main/binary-amd64/Packages  HttpError404

W: Failed to fetch https://launchpad.net/~alexlarsson/+archive/ubuntu/flatpak/dists/trusty/main/binary-i386/Packages  HttpError404

E: Some index files failed to download. They have been ignored, or old ones used instead
```

**Comment by elopio** [comment]
You can get ubuntu 16.04 in travis through LXD or docker.

**Comment by snaggen** [comment]
Don't have an Ubuntu around right now to test with, but it should just be
sudo add-apt-repository ppa:alexlarsson/flatpak 
$ sudo apt update 
$ sudo apt install flatpak

If that gives you a 404, it might be some problem with the canonical ppa platform.

**Comment by snaggen** [comment]
Maybe @alexlarsson can give some input about this.

**Comment by hoffmabc** [comment]
It’s just that one link with +archive that’s dead. Even with the ppa installed it does not have a flatpak package for Trusty.

**Comment by alexlarsson** [comment]
There are no trusty builds at the moment, only xenial and zesty. With some work it can probably be made to build on Trusty, but nobody has done that work.

**Comment by hoffmabc** [comment]
I can probably just use a docker instance to build it but our build process is already super tedious so I'll have to refactor the build process probably.

**Comment by arteniioleg** [comment]
I am on [Solus OS](https://solus-project.com/) and I need to run OpenBazaar **Client** to connect to my VPS which runs OpenBazaar Server.

Acceptable solutions:

* [snap](https://docs.snapcraft.io/core/install-solus)
* Docker image [for Client](https://hub.docker.com/u/openbazaar/). I will be able to start client GUI with [x11docker](https://github.com/mviereck/x11docker) as my current user and home directory mounted

**Comment by arteniioleg** [comment]
I think the simplest solution is to download and install the latest `.deb` on docker image build (instead of building the app from scratch).

**Comment by arteniioleg** [comment]
Check this out https://github.com/OpenBazaar/openbazaar-desktop/pull/1303

**Comment by jjeffryes** [comment]
This issue has been addressed sufficiently to be closed.

**Comment by arteniioleg** [comment]
https://github.com/arteniioleg/openbazaar-in-docker

**Comment by devasia2112** [comment]
I am trying to install with Arch Linux, not available on snap, I barely remember how I got it working years ago, still have the old backup but the project moved forward, I guess, the backup won't work. 
How can I install with Arch? 
Thanks, great bazaar!
